subprojects/core/src/main/java/org/gradle/api/internal/provider/DefaultListProperty.java

```diff
@@ -16,22 +16,62 @@
 
 package org.gradle.api.internal.provider;
 
+import com.google.common.base.Preconditions;
 import com.google.common.collect.ImmutableList;
 import org.gradle.api.Transformer;
 import org.gradle.api.provider.ListProperty;
 import org.gradle.api.provider.Provider;
 import org.gradle.internal.Cast;
 
 import javax.annotation.Nullable;
+import java.util.Collection;
 import java.util.List;
+import java.util.concurrent.Callable;
 
 class DefaultListProperty<T> implements PropertyInternal<List<T>>, ListProperty<T> {
     private static final Provider<ImmutableList<Object>> EMPTY_LIST = Providers.of(ImmutableList.of());
-    private Provider<? extends List<T>> provider = Cast.uncheckedCast(EMPTY_LIST);
+    private Provider<? extends Collection<T>> provider = Cast.uncheckedCast(EMPTY_LIST);
 
     public DefaultListProperty(Class<T> elementType) {
     }
 
+    @Override
+    public void add(final T element) {
+        addAll(Providers.of(ImmutableList.of(Preconditions.checkNotNull(element))));
+    }
+
+    @Override
+    public void add(final Provider<? extends T> providerOfElement) {
+        addAll(new DefaultProvider<List<T>>(new Callable<List<T>>() {
```

Comment:
this should be something like:

```
addAll(providerOfElement.map(new Transformer<List<T>, T>() { 
    List<T> transform(T t) { 
        return ImmutableList.of(t); 
    }
);
```

This way, no context from `providerOfElement` is lost.

Reply:
Done.